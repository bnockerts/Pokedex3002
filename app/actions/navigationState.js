import { PUSH_ROUTE, POP_ROUTE } from '../constants/actionTypes';

function navigatePush(state) {
  state = typeof state === 'string' ? { key: state, title: state } : state;
  return {
    type: PUSH_ROUTE,
    state
  };
}

function navigatePop() {
  return {
    type: POP_ROUTE,
  };
}

export default {
  navigatePush,
  navigatePop
};
