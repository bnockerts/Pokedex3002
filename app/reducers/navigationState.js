import { PUSH_ROUTE, POP_ROUTE } from '../constants/actionTypes';
import * as NavigationStateUtils from 'NavigationStateUtils'

export const initialState = {
  index: 0,
  routes: [
    {
      key: 'PokemonList',
      title: 'List of Pokemon'
    }
  ]
};

export default navigationState = (state = initialState, action) => {
  switch (action.type) {
    case PUSH_ROUTE:
      if (state.routes[state.index].key === (action.state && action.state.key)) {
        return state;
      }
		  return NavigationStateUtils.push(state, action.state);

    case POP_ROUTE:
      if (state.index === 0 || state.routes.length === 1) {
        return state;
      }
		  return NavigationStateUtils.pop(state);

    default:
      return state;
  }
};
