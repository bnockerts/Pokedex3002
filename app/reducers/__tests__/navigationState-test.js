import navigationState from '../navigationState';
import { initialState } from '../navigationState';
import navigationStateActions from '../../actions/navigationState';

it('handles unhandled action', () => {
  expect(navigationState(initialState, { type: '_NULL' }))
    .toMatchSnapshot();
});

it('handles PUSH_ROUTE action', () => {
  const testRouteName = 'testRoute';
  let expectedState = JSON.parse(JSON.stringify(initialState));
  expectedState.index = initialState.index + 1;
  expectedState.routes.push({ key: testRouteName, title: testRouteName });

  expect(navigationState(initialState, navigationStateActions.navigatePush(testRouteName)))
    .toEqual(expectedState);
});

it('handles POP_ROUTE action, pop on index 0', () => {
  expect(navigationState(initialState, navigationStateActions.navigatePop()))
    .toMatchSnapshot();
});

it('handles POP_ROUTE action, pop on index > 0', () => {
  const testRouteName = 'testRoute';
  const pushedState = navigationState(initialState, navigationStateActions.navigatePush(testRouteName));

  expect(navigationState(pushedState, navigationStateActions.navigatePop()))
    .toMatchSnapshot();
});
