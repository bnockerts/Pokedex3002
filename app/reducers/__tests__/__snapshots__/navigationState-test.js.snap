exports[`test handles POP_ROUTE action, pop on index > 0 1`] = `
Object {
  "index": 0,
  "routes": Array [
    Object {
      "key": "PokemonList",
      "title": "List of Pokemon",
    },
  ],
}
`;

exports[`test handles POP_ROUTE action, pop on index 0 1`] = `
Object {
  "index": 0,
  "routes": Array [
    Object {
      "key": "PokemonList",
      "title": "List of Pokemon",
    },
  ],
}
`;

exports[`test handles unhandled action 1`] = `
Object {
  "index": 0,
  "routes": Array [
    Object {
      "key": "PokemonList",
      "title": "List of Pokemon",
    },
  ],
}
`;
